### PR TITLE
FM2-636: Enhancements in Existing FHIR Task Resource

### PR DESCRIPTION
**Pull request by kalai-tw**
## Description of what I changed

Made below Changes to FHIR Task Entity Model:

- Added new columns to `fhir_task`:
  - `task_code`: to determine task type
  - `execution_start_time`: start time of execution
  - `execution_end_time`: end time of execution
  - `comment`: comment to task

- Added a new table `fhir_part_of_reference` to support subtasks to a parent task.

## Issue I worked on
see https://issues.openmrs.org/browse/FM2-636
talk thread ref https://talk.openmrs.org/t/contribution-to-fhir-task/42015

## Checklist: I completed these to help reviewers :)

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

- [x] I have **added tests** to cover my changes.

- [x] I ran `mvn clean package` right before creating this pull request 

- [x] All new and existing **tests passed**.

- [x] My pull request is **based on the latest changes** of the master branch.

### CODE REVIEW

File: api/src/test/java/org/openmrs/module/fhir2/api/translators/impl/TaskTranslatorImplTest.java [hunk 1]
```diff
@@ -946,6 +958,174 @@ public void toOpenmrsType_shouldTranslateMultipleInputValueTypes() {
...
+		
+		FhirTask task = new FhirTask();
+		Concept openmrsObject = new Concept();
+		CodeableConcept fhirObject = newCodeableConcept();
+		when(conceptTranslator.toFhirResource(openmrsObject)).thenReturn(fhirObject);
+		
+		task.setTaskCode(openmrsObject);
+		Task result = taskTranslator.toFhirResource(task);
+		assertThat(result.getCode(), CoreMatchers.notNullValue());
+		assertThat(result.getCode(), CoreMatchers.equalTo(fhirObject));
```
Comment (ibacher) [inline api/src/test/java/org/openmrs/module/fhir2/api/translators/impl/TaskTranslatorImplTest.java:973]:
```suggestion
		FhirTask task = new FhirTask();
		Concept openmrsObject = new Concept();
		CodeableConcept fhirObject = newCodeableConcept();
		when(conceptTranslator.toFhirResource(openmrsObject)).thenReturn(fhirObject);
		task.setTaskCode(openmrsObject);
		
		Task result = taskTranslator.toFhirResource(task);
		
		assertThat(result.getCode(), CoreMatchers.notNullValue());
		assertThat(result.getCode(), CoreMatchers.equalTo(fhirObject));
```

General rule here is to break things (visually) into: setup / execution / assertions. It makes the tests easier to tell what they do at a glance. This formatting should be adopted for all the tests in this file.

**Review by ibacher — CHANGES_REQUESTED** [review]
Thanks! This mostly LGTM. It would be good to add at least one integration test that verifies some of the changes here.

Also is there no need to be able to _search_ for tasks by these attributes? I would imagine, for example, that being able to search by code would be very useful.

File: api/src/test/java/org/openmrs/module/fhir2/api/translators/impl/TaskTranslatorImplTest.java [hunk 2]
```diff
@@ -946,6 +958,174 @@ public void toOpenmrsType_shouldTranslateMultipleInputValueTypes() {
...
+		executionPeriod.setEnd(endTime);
+		fhirTask.setExecutionPeriod(executionPeriod);
+		FhirTask result = taskTranslator.toOpenmrsType(fhirTask);
+		
+		assertThat(result.getExecutionStartTime(), equalTo(startTime));
+		assertThat(result.getExecutionEndTime(), equalTo(endTime));
+	}
+	
+	@Test
+	public void toOpenmrsType_shouldupdateExecutionTimeOnExistingTask() {
```
[the remark this answers is not in the document]
Comment (kalai-tw) [reply api/src/test/java/org/openmrs/module/fhir2/api/translators/impl/TaskTranslatorImplTest.java]: Fixed

**Comment by kalai-tw** [comment]
Hello @ibacher ! Thanksa for the feedbacks !

Yeap ! We would need Search By Code & Search By For Reference . Have added them in this new commit !

Also, Have Added the Integrations Tests now :)

**Comment by dkayiwa** [comment]
@kalai-tw do you mind looking into the above failing build?

**Review by ibacher — APPROVED** [review]

**Comment by ibacher** [comment]
@kalai-tw Could you rebase this branch on master, since we've committed the fix for the test?

**Comment by kalai-tw** [comment]
thanq @ibacher rebased & now the PR pipeline looks green ! can u help on squash & merging the same ?

**Pull request merged**